feat(auth): support Plex home profile login

0xSysR3ll opened the pull request:
<!--
    Please read contributing guide before submitting
    your pull request. Please fill in each section below to help us better prioritize your pull request. Thanks!
-->

#### Description

<!--- Describe your changes in detail -->
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

This PR adds support for Plex Home profiles.

### Authentication flow

```mermaid
flowchart TD
    A[User] -->|POST /auth/plex| B[Server]
    B -->|"single profile"| C["200 OK (session set)"]
    B -->|"multiple profiles"| D{{status: REQUIRES_PROFILE}}
    D --> E[Client shows profile selector]
    E -->|"POST /auth/plex/profile/select (profileId, pin?)"| F[Server]
    F -->|"needs PIN & none"| G{{status: REQUIRES_PIN}}
    F -->|"wrong PIN"| H[401 Invalid PIN]
    F -->|"imported profile"| I["200 OK (session set)"]
    F -->|"not imported & Allow new Plex logins OFF"| J[403 Access denied]
    F -->|"not imported & setting ON"| K[User created]
    K --> L[200 OK]

```

> [!NOTE]
> E‑mail fallback `<ownerPrefix>+<profile>@<ownerDomain>` (`plex.local` if the owner address has no domain)

P.S : Sorry for this "duplicate" PR (see #1503), but it's a complete rework.

## How Has This Been Tested?

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

#### Screenshot (if UI-related)

**Plex profile selector**
<img width="422" height="372" alt="image" src="https://github.com/user-attachments/assets/52a055e7-dbaf-4b5d-9bb3-52b1303cc4f0" />
**PIN entry**
<img width="418" height="409" alt="image" src="https://github.com/user-attachments/assets/e7680598-606b-4118-9673-663a442b25ea" />
**Invalid PIN**
<img width="455" height="519" alt="image" src="https://github.com/user-attachments/assets/58d7c54f-74dc-4db8-b291-aa99baba1df4" />
**Profile not imported (auto-import disabled)**
<img width="455" height="516" alt="image" src="https://github.com/user-attachments/assets/2a7f40e3-71b4-4fca-aba6-65768340b66d" />
**Users imported**
![image](https://github.com/user-attachments/assets/4c29d3c8-233f-4fae-8c18-5800ad91542f)

#### Plex Profile Import

In addition to the authentication flow, this PR now also adds support for importing Plex profiles from the admin interface:

- Admins can now import Plex profiles alongside regular Plex users
- Duplicate detection prevents importing profiles that conflict with existing users
- Clear visual indicators for potential duplicates
- Improved success and error messages that properly distinguish between users and profiles

##### Screenshots

**Plex Profile Import UI**
![image](https://github.com/user-attachments/assets/99d5a277-f3f8-4591-bff0-18b2f654cdad)
**Duplicate Detection**
![image](https://github.com/user-attachments/assets/78946701-fd46-4a88-98ff-a874c7bd199a)


#### To-Dos

- [ ] Make "Tautulli" feature work with plex profiles.
- [ ] Add "Linked account" support for plex profiles.
- [x] Support for plex users that are part of the Plex Home.
- [x] Add support for manual profiles import.
- [ ] Add support for profile's watch list. 

#### Issues Fixed or Closed

- Fixes #1402
- Closes #1503 

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I have read and followed the contribution [guidelines](https://github.com/seerr-team/seerr/blob/develop/CONTRIBUTING.md).
- [x] Disclosed any use of AI (see our [policy](https://github.com/seerr-team/seerr/blob/develop/CONTRIBUTING.md#ai-assistance-notice))
- [ ] I have updated the documentation accordingly.
- [ ] All new and existing tests passed.
- [x] Successful build `pnpm build`
- [x] Translation keys `pnpm i18n:extract`
- [x] Database migration (if required)


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Multi-profile Plex sign-in with profile selection UI and 4-digit PIN entry for protected profiles.
  * Import Plex Home profiles as separate users, with duplicate detection and separate import results.
  * New server endpoints to list/select Plex profiles; login responses may request profile selection or PIN.
  * DB changes to store Plex profile linkage on users.

* **Refactor**
  * Login flow reworked into a multi-step Plex authentication orchestration.

* **Documentation**
  * Added localization for profile selection, PIN entry, import, and duplicate messaging.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

0xSysR3ll committed:
feat(auth): support Plex home profile login

0xSysR3ll committed:
feat(auth): support Plex home profile login

0xSysR3ll committed:
fix: previous SSRF broke PIN verification

0xSysR3ll committed:
fix: previous SSRF broke PIN verification

0xSysR3ll commented on server/api/plextv.ts:
```diff
@@ -225,6 +266,144 @@
+    const urlPath = `/api/v2/home/users/${profileId}/switch`;
+
+    try {
+      const response = await axios.post(
+        urlPath,
+        { pin },
+        {
+          baseURL: 'https://clients.plex.tv',
+          headers: {
+            Accept: 'application/json',
+            'Content-Type': 'application/json',
+            'X-Plex-Token': this.authToken,
+            'X-Plex-Client-Identifier': randomUUID(),
+          },
+        }
+      );
```
Well, it's not an SSRF vulnerability anymore...
Host is hardcoded so no risks.
Could be a directory traversal though but limited Plex's API.
Can we silence this ?

gauthier-th replied:
Yes it can be ignored

0xSysR3ll replied:
Thanks !

0xSysR3ll committed:
fix(ui): autoFocus prop should not be used

0xSysR3ll committed:
fix(ui): autoFocus prop should not be used

0xSysR3ll committed:
fix: unused variables

0xSysR3ll committed:
fix: unused variables

0xSysR3ll committed:
fix: improve Plex profile selector with Next.js Image and safer null checks

0xSysR3ll committed:
fix: improve Plex profile selector with Next.js Image and safer null checks

0xSysR3ll committed:
fix: resolve TypeScript errors and ESLint warnings in login flow

- Add proper case block scoping with curly braces
- Update PlexProfile import path
- Add onError callback handling
- Fix dependency array in useEffect

0xSysR3ll committed:
fix: resolve TypeScript errors and ESLint warnings in login flow

- Add proper case block scoping with curly braces
- Update PlexProfile import path
- Add onError callback handling
- Fix dependency array in useEffect

0xSysR3ll committed:
fix(api): disable CodeQL warning for potential SSRF in Plex API call

0xSysR3ll committed:
fix(api): disable CodeQL warning for potential SSRF in Plex API call

0xSysR3ll committed:
fix(db): re-generate correct database migrations

0xSysR3ll committed:
fix(db): re-generate correct database migrations

0xSysR3ll committed:
fix: apply lint formatting

0xSysR3ll committed:
fix: apply lint formatting

0xSysR3ll committed:
fix(migration): update import statement to use type imports for TypeORM

0xSysR3ll committed:
fix(migration): update import statement to use type imports for TypeORM

0xSysR3ll committed:
fix(migration): add missing plex profiles fields and constraints

0xSysR3ll committed:
fix(migration): add missing plex profiles fields and constraints

0xSysR3ll committed:
fix(migration): remove unnecessary fields from Plex profiles migration

0xSysR3ll committed:
fix(migration): remove unnecessary fields from Plex profiles migration

0xSysR3ll commented:
I don't know if I did not RTFM properly but I struggled a bit with these data migration files sorry (I am using a docker env btw)...

gauthier-th commented:
> I don't know if I did not RTFM properly but I struggled a bit with these data migration files sorry (I am using a docker env btw)...

We have a step-by-step guide in our contribution guide to generate these migrations: https://github.com/fallenbagel/jellyseerr/blob/develop/CONTRIBUTING.md#migrations

gauthier-th commented:
> > > I don't know if I did not RTFM properly but I struggled a bit with these data migration files sorry (I am using a docker env btw)...
> > 
> > 
> > We have a step-by-step guide in our contribution guide to generate these migrations: https://github.com/fallenbagel/jellyseerr/blob/develop/CONTRIBUTING.md#migrations
> 
> Yep, I read this guide but I don't know why it generated unnecessary and incorrect things. Maybe I need to review my dev env.

Migrations for SQLite can be quite messy yes. IIRC SQLite doesn't support editing columns, hence the long migration files for SQLite.

0xSysR3ll committed:
fix(PlexPinEntry): remove duplicate error display

0xSysR3ll committed:
fix(PlexPinEntry): remove duplicate error display

0xSysR3ll committed:
fix(plexpin): remove duplicate error handling and reset pin on submit

0xSysR3ll committed:
fix(plexpin): remove duplicate error handling and reset pin on submit

0xSysR3ll committed:
feat: add support for importing Plex profiles

- Extend API endpoint to fetch and import Plex profiles
- Implement duplicate detection between users and profiles
- Add UI to display and select Plex profiles for import
- Improve error handling for duplicate records
- Enhance success/error messages with profile-specific text

0xSysR3ll committed:
feat: add support for importing Plex profiles

- Extend API endpoint to fetch and import Plex profiles
- Implement duplicate detection between users and profiles
- Add UI to display and select Plex profiles for import
- Improve error handling for duplicate records
- Enhance success/error messages with profile-specific text

0xSysR3ll committed:
fix: resolve Plex profile authentication and duplicate user issues

- Fix duplicate user creation when accessing profiles across accounts
- Prevent regular Plex users from being incorrectly marked as profiles
- Properly handle authentication for both Plex users and profiles
- Remove early returns in profile update code that broke authentication flow
- Add better uniqueness checks before creating profile users
- Ensure profile selection only shows to Plex account owner

0xSysR3ll committed:
fix: resolve Plex profile authentication and duplicate user issues

- Fix duplicate user creation when accessing profiles across accounts
- Prevent regular Plex users from being incorrectly marked as profiles
- Properly handle authentication for both Plex users and profiles
- Remove early returns in profile update code that broke authentication flow
- Add better uniqueness checks before creating profile users
- Ensure profile selection only shows to Plex account owner

0xSysR3ll committed:
fix(auth): streamline profile user creation and improve duplicate handling

0xSysR3ll committed:
fix(auth): streamline profile user creation and improve duplicate handling

0xSysR3ll committed:
fix(auth): resolve unused variable errors

0xSysR3ll committed:
fix(auth): resolve unused variable errors

0xSysR3ll commented:
Why does the cypress-run suddenly fails ? I didn't touch anything...
I am facing the same issue on my dev env.

0xSysR3ll committed:
fix(auth): remove unnecessary assignment of profileUser variable

0xSysR3ll committed:
fix(auth): remove unnecessary assignment of profileUser variable

fallenbagel commented:
> Why does the cypress-run suddenly fails ? I didn't touch anything...
> I am facing the same issue on my dev env.

Its not cypress that's failing. It's build that's failing because of a missing package

![Screenshot_20250427_005803_GitHub.jpg](https://github.com/user-attachments/assets/7f81f61f-a9ba-4e60-bd82-c8b764583689)

0xSysR3ll commented:
> > Why does the cypress-run suddenly fails ? I didn't touch anything...
> > I am facing the same issue on my dev env.
> 
> Its not cypress that's failing. It's build that's failing because of a missing type declaration/dep
> 
> ![Screenshot_20250427_005803_GitHub.jpg](https://github.com/user-attachments/assets/7f81f61f-a9ba-4e60-bd82-c8b764583689)
> 
> 

Mmm ok I will try to find the root cause but stack-trace is not that verbose...

0xSysR3ll committed:
fix(api): remove unused PlexProfile interface from plextv.ts

0xSysR3ll committed:
fix(api): remove unused PlexProfile interface from plextv.ts

0xSysR3ll committed:
fix: remove unused type imports and enhance data typing

0xSysR3ll committed:
fix: remove unused type imports and enhance data typing

0xSysR3ll commented:
I believe this PR is ready for review. I haven’t identified any issues so far, and there doesn’t appear to be any further work needed at this stage. Would it be appropriate to provide a tag for testing, or is it still too early ?

fallenbagel commented:
> I believe this PR is ready for review. I haven’t identified any issues so far, and there doesn’t appear to be any further work needed at this stage. Would it be appropriate to provide a tag for testing, or is it still too early ?

I can push a preview tag tonight

0xSysR3ll commented:
> > I believe this PR is ready for review. I haven’t identified any issues so far, and there doesn’t appear to be any further work needed at this stage. Would it be appropriate to provide a tag for testing, or is it still too early ?
> 
> I can push a preview tag tonight

Awesome 🔥

edbourque0 commented:
I'm not sure if this is the right place to ask this question (sorry if it's not) but does this pull request add the ability for each profile to request movies in jellyseerr thru their plex watchilsts ? This feature is already available but I was just wondering if this would add a "per profile" request thru each profile's watchlist.

Thank you!

0xSysR3ll commented:
> I'm not sure if this is the right place to ask this question (sorry if it's not) but does this pull request add the ability for each profile to request movies in jellyseerr thru their plex watchilsts ? This feature is already available but I was just wondering if this would add a "per profile" request thru each profile's watchlist.
> 
> 
> 
> Thank you!

I haven’t had the chance to review or test the watchlist feature yet, so I can’t confirm at the moment. However, if it doesn’t currently support this functionality, it would be great to consider integrating it.

0xSysR3ll committed:
fix(auth): critical auth vulnerability in Plex profile selection

This fixes a critical security vulnerability where non-imported profiles could
authenticate as arbitrary users if their profile IDs shared any substring.

0xSysR3ll committed:
fix(auth): critical auth vulnerability in Plex profile selection

This fixes a critical security vulnerability where non-imported profiles could
authenticate as arbitrary users if their profile IDs shared any substring.

edbourque0 commented:
> > I believe this PR is ready for review. I haven’t identified any issues so far, and there doesn’t appear to be any further work needed at this stage. Would it be appropriate to provide a tag for testing, or is it still too early ?
> 
> I can push a preview tag tonight

What is the name of the tag ? I'm eager to test this

0xSysR3ll commented:
> > > I believe this PR is ready for review. I haven’t identified any issues so far, and there doesn’t appear to be any further work needed at this stage. Would it be appropriate to provide a tag for testing, or is it still too early ?
> 
> > 
> 
> > I can push a preview tag tonight
> 
> 
> 
> What is the name of the tag ? I'm eager to test this

I think fallenbagel forgot.
Must be busy.

fallenbagel commented:
> > > > I believe this PR is ready for review. I haven’t identified any issues so far, and there doesn’t appear to be any further work needed at this stage. Would it be appropriate to provide a tag for testing, or is it still too early ?
> > 
> > > 
> > 
> > > I can push a preview tag tonight
> > 
> > 
> > 
> > What is the name of the tag ? I'm eager to test this
> 
> I think fallenbagel forgot.
> Must be busy.

Yeah sorry I got busy 😅
Let me push it

fallenbagel commented:
Preview should be available in half an hour as:
`preview-plex-home-profile`

fallenbagel commented:
Please use the discussion #1627 for discussions related to this PR

edbourque0 commented:
> @edbourque0 Hello, sorry to ping you here but didn't find anywhere else to do it. Did you manage to find some time to test ?

Sorry just got back from vacation. I'll be taking a look at this during the weekend and will be posting if there are any issues.

0xSysR3ll commented:
> Here are the observations I have made while testing : 
> 
> 1. The first time a user logs in, the user is created but it is not authenticated (redirected in the Jellyseerr app). I need to login as a profile that has already been created, logout and then I can sign-in with the profile I previously tried with.
> 
> 
> 
> 2. The watchlist auto-request feature does not work with account other than the Plex admin. I personally think this would be a great feature to add to this PR.
> 
> 
> 
> 3. The pin input bos should be autoselected when a user has a code and since it is always a 4 digit code, why not submit as soon as 4 digits are entered.
> 
> 
> 
> Bonus : Wait for the OpenID integration (#1505) to be done so we can have both features.
> 
> 
> 
> Other than this I have found no bugs. @0xSysR3ll Thank you for making this possible!

Thanks a lot for your precious review @edbourque0 !
Would you just mind redact this in the [conversation](#1627) instead ? As we can create threads - will be easier to follow :)

0xSysR3ll committed:
fix(auth): set session for newly created user after Plex profile selection

0xSysR3ll committed:
fix(auth): set session for newly created user after Plex profile selection

0xSysR3ll committed:
fix(auth): enhance PlexPinEntry component with auto-focus and auto-submit features

0xSysR3ll committed:
fix(auth): enhance PlexPinEntry component with auto-focus and auto-submit features

0xSysR3ll committed:
fix: resolve TypeScript error in PlexPinEntry onClick handler

0xSysR3ll committed:
fix: resolve TypeScript error in PlexPinEntry onClick handler

0xSysR3ll committed:
feat(plex-auth): enhanced layout and styling for better UX

0xSysR3ll committed:
feat(plex-auth): enhanced layout and styling for better UX

0xSysR3ll committed:
feat(auth): add handling for existing profile users during Plex login

0xSysR3ll committed:
feat(auth): add handling for existing profile users during Plex login

0xSysR3ll committed:
refactor(api): rename and enhance profile switching logic

0xSysR3ll committed:
refactor(api): rename and enhance profile switching logic

0xSysR3ll committed:
feat(login): add back navigation functionality to PlexProfileSelector

0xSysR3ll committed:
feat(login): add back navigation functionality to PlexProfileSelector

0xSysR3ll committed:
feat(plex): add numericId to enhance Plex profile support

0xSysR3ll committed:
feat(plex): add numericId to enhance Plex profile support

0xSysR3ll committed:
fix: use correct Plex profile ID for Tautulli watch data

Now profile users see their own watch history instead of the main account's data.

0xSysR3ll committed:
fix: use correct Plex profile ID for Tautulli watch data

Now profile users see their own watch history instead of the main account's data.

0xSysR3ll committed:
refactor: replace isPlexProfile boolean with UserType.PLEX_PROFILE enum

0xSysR3ll committed:
refactor: replace isPlexProfile boolean with UserType.PLEX_PROFILE enum

0xSysR3ll committed:
feat: add Plex Profile badges in user list and profile settings

0xSysR3ll committed:
feat: add Plex Profile badges in user list and profile settings

0xSysR3ll committed:
feat(i18n): add missing translation for Plex Profile

0xSysR3ll committed:
feat(i18n): add missing translation for Plex Profile

0xSysR3ll committed:
fix(auth): remove duplicate userType assignment

0xSysR3ll committed:
fix(auth): remove duplicate userType assignment

0xSysR3ll committed:
fix: correct self-closing tags for span elements

0xSysR3ll committed:
fix: correct self-closing tags for span elements

0xSysR3ll committed:
fix(auth): protect plex profiles retrieval endpoint

0xSysR3ll committed:
fix(auth): protect plex profiles retrieval endpoint

0xSysR3ll committed:
feat(migration): add Plex profiles support for PostgreSQL and SQLite

0xSysR3ll committed:
feat(migration): add Plex profiles support for PostgreSQL and SQLite

0xSysR3ll committed:
feat(i18n): extract translations

0xSysR3ll committed:
feat(i18n): extract translations

0xSysR3ll committed:
feat(api): add required fields for user and profile objects in API schema

0xSysR3ll committed:
feat(api): add required fields for user and profile objects in API schema

0xSysR3ll committed:
fix(api): change Accept header from JSON to XML for Plex API requests

0xSysR3ll committed:
fix(api): change Accept header from JSON to XML for Plex API requests

0xSysR3ll committed:
fix(auth): validate Plex authToken against profile before session binding

0xSysR3ll committed:
fix(auth): validate Plex authToken against profile before session binding

0xSysR3ll committed:
fix(auth): add error handling for loading Plex profiles during login

0xSysR3ll committed:
fix(auth): add error handling for loading Plex profiles during login

0xSysR3ll committed:
fix(auth): clear PIN error on invalid profile selection during login

0xSysR3ll committed:
fix(auth): clear PIN error on invalid profile selection during login

0xSysR3ll committed:
refactor(user): remove unnecessary user check

0xSysR3ll committed:
refactor(user): remove unnecessary user check

0xSysR3ll committed:
chore: apply linting

0xSysR3ll committed:
chore: apply linting

0xSysR3ll committed:
fix(auth): use UUID instead of string conversion

0xSysR3ll committed:
fix(auth): use UUID instead of string conversion

0xSysR3ll committed:
fix(auth): add authorization checks for user profile access

0xSysR3ll committed:
fix(auth): add authorization checks for user profile access

0xSysR3ll committed:
fix(auth): ensure processing state is reset after login error handling

0xSysR3ll committed:
fix(auth): ensure processing state is reset after login error handling

0xSysR3ll committed:
fix(api): increase timeout for Plex API requests

0xSysR3ll committed:
fix(api): increase timeout for Plex API requests

0xSysR3ll committed:
fix(api): add 403 error response for unauthorized user profile access

0xSysR3ll committed:
fix(api): add 403 error response for unauthorized user profile access

0xSysR3ll committed:
fix(auth): update profile selection logic to return main user ID only when multiple profiles exist

0xSysR3ll committed:
fix(auth): update profile selection logic to return main user ID only when multiple profiles exist

0xSysR3ll committed:
fix(settings): remove redundant user type check

0xSysR3ll committed:
fix(settings): remove redundant user type check

0xSysR3ll committed:
refactor(auth): remove profileId prop from PlexPinEntry component

0xSysR3ll committed:
refactor(auth): remove profileId prop from PlexPinEntry component

0xSysR3ll committed:
refactor(auth): remove unused mainUserId and authToken props from PlexProfileSelector

0xSysR3ll committed:
refactor(auth): remove unused mainUserId and authToken props from PlexProfileSelector

0xSysR3ll committed:
refactor(auth): remove unused profilesRef reference in Login component

0xSysR3ll committed:
refactor(auth): remove unused profilesRef reference in Login component

0xSysR3ll committed:
fix(auth): validate mainUserId and authToken in Plex profile selection

0xSysR3ll committed:
fix(auth): validate mainUserId and authToken in Plex profile selection

0xSysR3ll committed:
fix(auth): query only by the unique plexProfileId

0xSysR3ll committed:
fix(auth): query only by the unique plexProfileId

0xSysR3ll committed:
feat(PlexImportModal): missing localized messages

0xSysR3ll committed:
feat(PlexImportModal): missing localized messages

0xSysR3ll committed:
refactor(auth): simplify conditional user check

0xSysR3ll committed:
refactor(auth): simplify conditional user check

0xSysR3ll committed:
feat(i18n): add new localized messages

0xSysR3ll committed:
feat(i18n): add new localized messages

0xSysR3ll committed:
fix(PlexProfileSelector): make profile selection handler asynchronous

0xSysR3ll committed:
fix(PlexProfileSelector): make profile selection handler asynchronous

0xSysR3ll committed:
feat(api): add validation for optional 4-digit profile PIN

0xSysR3ll committed:
feat(api): add validation for optional 4-digit profile PIN

0xSysR3ll committed:
refactor(Login): streamline PIN profile data retrieval

0xSysR3ll committed:
refactor(Login): streamline PIN profile data retrieval

0xSysR3ll committed:
refactor(auth): remove redundant user validation

0xSysR3ll committed:
refactor(auth): remove redundant user validation

0xSysR3ll committed:
refactor(auth): remove redundant exact ID match check

0xSysR3ll committed:
refactor(auth): remove redundant exact ID match check

0xSysR3ll committed:
refactor(plextv): remove useless try-catch block

0xSysR3ll committed:
refactor(plextv): remove useless try-catch block

0xSysR3ll committed:
refactor(PlexImportModal): enhance type definitions

0xSysR3ll committed:
refactor(PlexImportModal): enhance type definitions

0xSysR3ll committed:
refactor(auth): enforce PIN requirement for protected main accounts

0xSysR3ll committed:
refactor(auth): enforce PIN requirement for protected main accounts

0xSysR3ll committed:
refactor(plextv): update error handling

0xSysR3ll committed:
refactor(plextv): update error handling

0xSysR3ll committed:
fix(auth): require profile login via profile/select to enforce PIN

0xSysR3ll committed:
fix(auth): require profile login via profile/select to enforce PIN
